### ISSUE TITLE
Allow open blocks to be nested

### ISSUE THREAD

**Issue by bimlas** [asciidoctor/asciidoctor#1121]
I'm using _asciidoctor_ to replace a website which is made in a _wysiwyg html editor_  to be responsive. I have to use _openblocks_ frequently and it may be cool to nest those in to each other. I think a signing character can be solve this, for example:

```
[.parent]
--A
[.child]
--B
content
--B
[.another_child]
--B
content
--B
--A
```

Wdyt?

To know what i mean, check the [old site](http://www.szizaltrend.hu/nadfonatok.html) and the [new](http://www.ro.naturtrend.com/ratanandthonet.html) _asciidoctor_ version. I need nested `div`s to vertically align the tables on the left and the text on the right side, like this:

```
+--------------------+
|Container   +------+|
|+----------+|Text  ||
||Table div ||div   ||
|+----------+|      ||
|            +------+|
+--------------------+
```

**Comment by esammer** [comment]
@mojavelinux Isn't it possible to just have the parser count open/close pairs (since you can't have open A, open B, close A, close B anyway)?

**Comment by mojavelinux** [comment]
Yes, that's a possibility, but I think it makes it confusing for the human. We always have to strike a balance the two perspectives with content. I'm not ruling out the idea, but it does make me a bit nervous.

**Comment by mojavelinux** [comment]
We could also consider using tilde's if we are worried about backwards-compatibility.

```
~~
outer
~~~~
inner
~~~~
outer
~~
```

**Comment by mojavelinux** [comment]
I like the idea of tilde's, but also making the minimum number of characters 4 instead of 2, thus bringing open blocks in alignment with all other structural block containers (aka delimited blocks).

**Comment by mojavelinux** [comment]
The more I write in AsciiDoc, the more I find that I need this. I want to switch from (and deprecate) the two hyphen syntax and switch to four or more tildes. It just makes sense.

**Comment by mojavelinux** [comment]
In the initial proposal, I've decide to only allow this syntax for contexts that don't already have designated delimiters. Therefore, it would not be possible to define a sidebar this way. I think that makes the blocks harder to parse by editors.

See #3010.

**Comment by mojavelinux** [comment]
After seeking advice from several people about the timing of this change, I've decided to defer this until after the spec is underway. It sends a mixed message to make a prominent syntax change while also saying we want to route changes to the language through a working group. The change will still have to go into the reference implementation before it can be made part of the spec, but it's premature to do it before we've even discussed it. And it also happens to be a good candidate for the first group of topics to discuss.

The only syntax changes that we'll put into 2.0.0 at this point are ones that pertain to existing syntax.

**Comment by ctron** [comment]
I just ran into an issue with this too. I think it is a shame, that a generic component such as this didn't get any attention, and that this is an open issue for more than 5 years.

My use case is a generator, which generates documentation in asciidoc, that gets rendered with Antora. Now with generators it isn't a human creating the content, so the esthetics of the actual asciidoc aren't a real concern. What is more of a problem, is that the generator needs to understand the different nesting levels and continuations when templates get expanded. At some point, you just don't know that. Having a wrapping "open block" could easily solve that problem. But as they cannot be nested, they are far less powerful than they could be.

And while there are all kinds of tweak for other blocks, it seems like there is nothing in the context of "open blocks".

So, why not go with the proposal from 2015 and use tilde?

**Comment by mojavelinux** [comment]
This is planned for the first version of the AsciiDoc specification. Once it is ratified there, I will merge it in Asciidoctor. We get that it's important. But we also have to honor the process that we have established.

**Comment by ctron** [comment]
Thanks for the update, that really helps. Is there any timeline for this change?

**Comment by aisbergde** [comment]
Is there any current solution or work around to use nested open blocks?

In the current docs only the pair of 2 '--' is documented. Nothing about 4+ tilde: issue #3010 is deferred. This was 3 years ago.

Some issues regarding nested open blocks can be found, some of them are closed and link to other issues. #2340 #1121 #3010

I understand that a solution would be fine, which will not break the syntax in existing documents. But we need any solution. It is hard to generate automated content for documentations without nested open blocks. It is maybe fine for manually generated content without nested open blocks.

Here are some nice ideas. from 2014. Now we have 2022.

Maybe we should or could discuss on Zulip chat?

**Comment by veselov** [comment]
This is quite a nuisance for automatically generated docs, where the generator just needs to encompass a compound content without any formatting change.
I understand that changing the block delimiter syntax is major spec breaking change, but can it just be allowed to have blocks masquerade as open? 
Looking at https://docs.asciidoctor.org/asciidoc/latest/blocks/masquerading/, there are these rules defining which blocks can masquerade as which other blocks. Nothing can be masqueraded as an open block.
All that it would take to support nested open block is to at least allow that masquerading. I'm also curious as to the rule of the masquerading matrix, i.e., what is a reason for not allowing complete freedom in masquerading.

This is what I'm trying to do:
```asciidoc
DD::
+
[open]
====

before nested

ANOTHER DD:::
+
[open]
=====
nested
=====

after nested

====
```

I can't find any way to use any other block delimiters, and keep the style as normal.

**Comment by mojavelinux** [comment]
I have good news for you. First, your proposed workaround is already possible using a very simple extension. I have pushed that extension to the lab. See:

https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/main/lib/nested-open-block.rb

Second, as for the status of the issue at hand, there's news there too. We established the policy that once an SDR is created in the AsciiDoc Language project for a feature, we are prepared to add it to Antora. And, in fact, there is an SDR for nested open blocks. See https://gitlab.eclipse.org/eclipse/asciidoc-lang/asciidoc-lang/-/blob/main/spec/sdrs/sdr-001-open-block-delimiter.adoc?ref_type=heads So we'll go ahead include this new delimited block in Asciidoctor 2.1.0.

**Comment by veselov** [comment]
That's very helpful, thank you.
However, we are using Antora, and AFAICU, we are bound to JS extensions only.
I figure to enable this, we'll need to port the nested open block Ruby extension to JS before we can utilize it...
Thank you!

**Comment by mojavelinux** [comment]
> However, we are using Antora, and AFAICU, we are bound to JS extensions only.

The extension can also be written for Asciidoctor.js so it can be used with Antora. It's really just one line of code after the configuration. If you need help with that, please ask in the Antora chat.

**Comment by dahopem** [comment]
@mojavelinux, could you be so kind and rebase (or re-implement) your branch [issue-1121-tilde-open-block](https://github.com/mojavelinux/asciidoctor/tree/issue-1121-tilde-open-block) used in your pull request #3010?

Alternatively: Is there any AsciiDoctor 2.1 branch which already contains an implementation?

This would considerably lower the barrier to use^Wtest the solution (4 or more "~" characters) which has been [proposed already 11 months ago as the new standard for open blocks](https://gitlab.eclipse.org/eclipse/asciidoc-lang/asciidoc-lang/-/merge_requests/9), as just a slightly different ```gem 'asciidoctor'``` line would be needed in the respective ```Gemfile``` then.

Alternatively: What is the next barrier to getting this issue resolved?

**Comment by mojavelinux** [comment]
I will implement it when I have the time. I don't have the time right now.

**Comment by mojavelinux** [comment]
One key part of this change is that we're going to do away with an early design mistake of this language, which is to no longer allow an open block defined using this new delimiter to masquerade as any other type of block (except the special abstract and partintro blocks, which are really just enclosures). In other words, this new open block cannot be made into an example block or listing block, etc*. An open block is an open block. It can have it's own role, of course, and extensions can still change the context. But the core language won't support it. Otherwise, it's extremely difficult for tooling.

\* The two hyphen delimiter will retain the existing functionality for backwards compatibility, but the spec may not allow it.
